In [8]:
# from lightfm.data import Dataset
import pandas as pd
import numpy as np
from scipy.sparse import coo_matrix
from lightfm import LightFM
from lightfm.data import Dataset


lastfm_api_key = "97d5a64d5ba4a8bc580b752ceff3b87f"
lastfm_secret = "35175090bd61f6f16ac607bd26e5b1de"

In [7]:
pip install lightfm

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 316.4/316.4 kB 2.7 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Created wheel for lightfm: filename=lightfm-1.17-cp310-cp310-linux_x86_64.whl size=808331 sha256=15c792cff9304e12b42dacd03c601e9eeae88325bbc1ebeda699fce194831f1f
  Stored in directory: /root/.cache/pip/wheels/4f/9b/7e/0b256f2168511d8fa4dae4fae0200fdbd729eb424a912ad636
Successfully built lightfm


In [2]:
import pandas as pd
import requests
from collections import deque
'''
def get_lastfm_friends_bfs(start_username, api_key, min_users=5000):
    discovered = set([start_username])
    queue = deque([start_username])
    collected_friends = []

    while queue and len(collected_friends) < min_users:
        current_user = queue.popleft()
        url = f"http://ws.audioscrobbler.com/2.0/?method=user.getfriends&user={current_user}&api_key={api_key}&format=json"

        try:
            response = requests.get(url)
            data = response.json()

            if 'error' in data:
                print(f"Error fetching data for user {current_user}: {data['message']}")
                continue

            users = data.get('friends', {}).get('user', [])
            for user in users:
                friend_name = user['name']
                if friend_name not in discovered:
                    discovered.add(friend_name)
                    queue.append(friend_name)
                    collected_friends.append(friend_name)
                    if len(collected_friends) >= min_users:
                        break

            print(f"Collected {len(collected_friends)} friends so far...")
        except Exception as e:
            print(f"An error occurred while processing user {current_user}: {e}")

    return collected_friends[:min_users]

def get_top_tracks_for_users(users, api_key):
    user_tracks = []
    user_artists = []
    user_playcounts = []
    user_ids = []

    for user in users:
        result = get_top_tracks(user, api_key)
        tracks, artists, playcounts = [], [], []

        for item in result['toptracks']['track'][:10]:
            tracks.append(item['name'])
            artists.append(item['artist']['name'])
            playcounts.append(item['playcount'])

        user_tracks.append(tracks)
        user_artists.append(artists)
        user_playcounts.append(playcounts)
        user_ids.append(user)

    df = pd.DataFrame({
        'User': user_ids,
        'Tracks': user_tracks,
        'Artists': user_artists,
        'Playcounts': user_playcounts
    })

    return df

def get_top_tracks(user, api_key):
    url = f"http://ws.audioscrobbler.com/2.0/?method=user.gettoptracks&user={user}&api_key={api_key}&format=json"
    response = requests.get(url)
    result = response.json()
    return result

start_username = "Bans77"

users = get_lastfm_friends_bfs(start_username, lastfm_api_key, min_users=100)

df_top_tracks = get_top_tracks_for_users(users, lastfm_api_key)

print(df_top_tracks.head())
'''


'\ndef get_lastfm_friends_bfs(start_username, api_key, min_users=5000):\n    discovered = set([start_username]) \n    queue = deque([start_username])     \n    collected_friends = []              \n\n    while queue and len(collected_friends) < min_users:\n        current_user = queue.popleft()\n        url = f"http://ws.audioscrobbler.com/2.0/?method=user.getfriends&user={current_user}&api_key={api_key}&format=json"\n        \n        try:\n            response = requests.get(url)\n            data = response.json()\n            \n            if \'error\' in data:\n                print(f"Error fetching data for user {current_user}: {data[\'message\']}")\n                continue\n            \n            users = data.get(\'friends\', {}).get(\'user\', [])\n            for user in users:\n                friend_name = user[\'name\']\n                if friend_name not in discovered:\n                    discovered.add(friend_name)\n                    queue.append(friend_name)\n     

In [3]:
from google.colab import drive
drive.mount('/content/drive')

df = pd.read_csv('/content/drive/MyDrive/Googlecolab/user_songs_filtered.csv')
df

Mounted at /content/drive


Username         country  registered_year  track_count  \
0               emosoup   United States             2015        13520   
1       maiconslavieiro          Brazil             2019        18795   
2             velenious   United States             2013        32367   
3           Antimemetic   United States             2023         4030   
4         frankcreature  Czech Republic             2023         2528   
...                 ...             ...              ...          ...   
393115           Txxinn  United Kingdom             2020        29495   
393116        cow_puppy          Canada             2021        32105   
393117        cow_puppy          Canada             2021        32105   
393118        cow_puppy          Canada             2021        32105   
393119        cow_puppy          Canada             2021        32105   

        artist_count                          track_name  \
0               3386                              Higher   
1               6513                              Higher   
2               8810                              Higher   
3                711                              Higher   
4                706                              Higher   
...              ...                                 ...   
393115          8086                               candy   
393116          2251  Fallen Tree Bridge (Brave Rainbow)   
393117          2251                  The Breathing Fire   
393118          2251       Solar Mirror Anthology Vol. 6   
393119          2251           Outlook Remains Untouched   

                 artist_name  rank  playcount  \
0                Sleep Token     1       1321   
1                Sleep Token    21        151   
2                Sleep Token    19       1259   
3                Sleep Token    32         29   
4                Sleep Token    43         43   
...                      ...   ...        ...   
393115       guccihighwaters    50        334   
393116           Candy Claws    33         39   
393117           Candy Claws    43         35   
393118        Sound of Ceres    46         33   
393119  ||||||||||||||||||||    48         31   

                                                   lyrics  ...  emotion1  \
0       you say you wont begin again\ncapitulate and l...  ...  negative   
1       you say you wont begin again\ncapitulate and l...  ...  negative   
2       you say you wont begin again\ncapitulate and l...  ...  negative   
3       you say you wont begin again\ncapitulate and l...  ...  negative   
4       you say you wont begin again\ncapitulate and l...  ...  negative   
...                                                   ...  ...       ...   
393115  i might have to block you know its for the bet...  ...  negative   
393116  \nfall asleep up high on moss\ncovered fallen ...  ...  positive   
393117  to halt the freeze of night\ngreen canopy brig...  ...  positive   
393118  \n\nthe king asked her and what was their stor...  ...  positive   
393119  \nhomeostasis refers to the bodys automatic ef...  ...  positive   

        emotion1_score  emotion2 emotion2_score  \
0             0.227031   sadness       0.156834   
1             0.227031   sadness       0.156834   
2             0.227031   sadness       0.156834   
3             0.227031   sadness       0.156834   
4             0.227031   sadness       0.156834   
...                ...       ...            ...   
393115        0.142786  surprise       0.062549   
393116        0.258182  negative       0.047825   
393117        0.180994      fear       0.139447   
393118        0.179915  negative       0.123571   
393119        0.267113  negative       0.158076   

                                                     mfcc  \
0       [-238.95233    173.60838    -18.518389    40.3...   
1       [-238.95233    173.60838    -18.518389    40.3...   
2       [-238.95233    173.60838    -18.518389    40.3...   
3       [-238.95233    173.60838    -18.518389    40.3... 

In [4]:
df_top_tracks = df

# Group by 'Username' and aggregate the other columns into lists
aggregated_data = df_top_tracks.groupby('Username').agg({
    'track_name': list,
    'artist_name': list,
    'playcount': list
}).reset_index()

'''
# You might want to also aggregate other columns like 'country', 'registered_year', etc., if needed
# e.g., taking the first or most common value within each group
aggregated_data['country'] = df_top_tracks.groupby('Username')['country'].agg(lambda x: x.mode()[0])
aggregated_data['registered_year'] = df_top_tracks.groupby('Username')['registered_year'].agg(lambda x: x.mode()[0])
aggregated_data['listeners'] = df_top_tracks.groupby('track_name')['listeners'].agg(lambda x: x.mode()[0])
aggregated_data['listeners'] = df_top_tracks.groupby('track_name')['listeners'].agg(lambda x: x.mode()[0])
aggregated_data['total_playcount'] = df_top_tracks.groupby('track_name')['total_playcount'].agg(lambda x: x.mode()[0])
aggregated_data['toptags'] = df_top_tracks.groupby('track_name')['toptags'].agg(lambda x: x.mode()[0])
'''

# Rename the columns to match the second image
aggregated_data.rename(columns={
    'Username': 'User',
    'track_name': 'Tracks',
    'artist_name': 'Artists',
    'playcount': 'Playcounts'
}, inplace=True)

# The resulting DataFrame should be in the desired format
aggregated_data


User                                             Tracks  \
0        -Dolorosa-  [I Will Follow You Into The Dark, Destruction ...   
1        -itssoeasy  [Drive, Dig, Holding Someone's Hair Back, Anna...   
2             0-172  [Head Hunter, Pachuca Sunrise, Throwin' Shapes...   
3          23linear  [8000, Crystalline, Neurosomatic Circuit, Proc...   
4     40belowsummer  [My Own Summer (Shove It), One Step Closer, In...   
...             ...                                                ...   
9478           zuoa  [Be Quiet and Drive (Far Away), Girls just wan...   
9479          zurvk  [Girls just want to have fun, Western Union, L...   
9480         zuzu80  [LEFT RIGHT, OMG, Ditto, Yes - Remastered, Fas...   
9481       zxHelios  [i see everything, fax or nah?? (like fr fr), ...   
9482    zzakkkkkkkk  [the Glow, pt. 2, Jesus, Etc., Nude, All I Nee...   

                                                Artists  \
0     [Death Cab for Cutie, Chelsea Wolfe, Have a Ni...   
1     [Incubus, Incubus, Circa Survive, Incubus, Kil...   
2     [Dance Gavin Dance, Minus the Bear, Minus the ...   
3     [Extrawelt, Younger Brother, Androcell, Androc...   
4     [Deftones, Linkin Park, Linkin Park, Faith No ...   
...                                                 ...   
9478  [Deftones, Bladee, blink-182, Bladee, Bladee, ...   
9479  [Bladee, Ecco2K, Cartier God, p4rkr, p4rkr, p4...   
9480  [XG, NewJeans, NewJeans, Manic Street Preacher...   
9481  [93FEETOFSMOKE, 93FEETOFSMOKE, 93FEETOFSMOKE, ...   
9482  [The Microphones, Wilco, Radiohead, Radiohead,...   

                                             Playcounts  
0     [283, 225, 187, 190, 263, 302, 174, 327, 173, ...  
1     [62, 42, 23, 45, 23, 17, 44, 20, 23, 25, 55, 1...  
2     [154, 185, 166, 144, 195, 165, 193, 185, 180, ...  
3     [68, 97, 73, 77, 324, 262, 142, 120, 118, 112,...  
4     [166, 130, 155, 134, 152, 169, 131, 150, 142, ...  
...                                                 ...  
9478  [108, 153, 91, 107, 91, 115, 106, 121, 110, 11...  
9479  [127, 119, 127, 310, 146, 138, 111, 133, 165, ...  
9480  [36, 37, 37, 44, 46, 36, 44, 48, 44, 225, 75, ...  
9481  [226, 675, 260, 297, 346, 259, 232, 315, 509, ...  
9482  [47, 59, 68, 63, 70, 61, 50, 48, 54, 54, 46, 6...  

[9483 rows x 4 columns]

In [9]:
records = []
for i, row in aggregated_data.iterrows():
    user = row['User']
    for track, artist, playcount in zip(row['Tracks'], row['Artists'], row['Playcounts']):
        track_artist = f"{track} - {artist}"
        records.append((user, track_artist, playcount))

df_flat = pd.DataFrame(records, columns=['User', 'Track_Artist', 'Playcount'])
print(df_flat)
dataset = Dataset()
dataset.fit(users=df_flat['User'].unique(),
            items=df_flat['Track_Artist'].unique())

(interactions, weights) = dataset.build_interactions([(x['User'], x['Track_Artist'], float(x['Playcount'])) for index, x in df_flat.iterrows()])

               User                                       Track_Artist  \
0        -Dolorosa-  I Will Follow You Into The Dark - Death Cab fo...   
1        -Dolorosa-  Destruction Makes the World Burn Brighter - Ch...   
2        -Dolorosa-                       Bloodhail - Have a Nice Life   
3        -Dolorosa-                      Such Small Hands - La Dispute   
4        -Dolorosa-                          Teardrop - Massive Attack   
...             ...                                                ...   
393115  zzakkkkkkkk                             Anne - John Frusciante   
393116  zzakkkkkkkk            Passover - 2007 Remaster - Joy Division   
393117  zzakkkkkkkk                             Get the Dutch - Duster   
393118  zzakkkkkkkk                             Hope - John Frusciante   
393119  zzakkkkkkkk                         The Real - John Frusciante   

        Playcount  
0             283  
1             225  
2             187  
3             190  
4          

In [11]:
pip install implicit

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 8.9/8.9 MB 17.5 MB/s eta 0:00:00


In [12]:
from scipy.sparse import coo_matrix
from implicit.als import AlternatingLeastSquares

df_flat['user_id_code'] = df_flat['User'].astype("category").cat.codes
df_flat['item_id_code'] = df_flat['Track_Artist'].astype("category").cat.codes

user_item_matrix = coo_matrix((df_flat['Playcount'].astype(np.float32),
                                (df_flat['user_id_code'], df_flat['item_id_code'])))

user_item_matrix_csr = user_item_matrix.tocsr()

item_user_matrix = user_item_matrix.T.tocsr()

In [13]:
from implicit.evaluation import train_test_split
train_interaction_matrix, test_interaction_matrix = train_test_split(interactions, train_percentage=0.8 )

# AlternatingLeastSquares
The AlternatingLeastSquares model from the implicit library primarily focuses on matrix factorization techniques that work directly with user-item interaction data. It doesn't natively support incorporating user or item features directly into the model during the matrix factorization process in the way that models like LightFM do, which are designed as hybrid recommendation models capable of utilizing both interaction data and metadata (e.g., genre, user demographics).

In [148]:
model_implicit = AlternatingLeastSquares(factors=500, iterations=50, calculate_training_loss=True)
model_implicit.fit(train_interaction_matrix)

user_id_map = dict(zip(df_flat['User'].astype("category"), df_flat['user_id_code']))
item_id_map = dict(zip(df_flat['Track_Artist'].astype("category"), df_flat['item_id_code']))

user_code_to_id_map = {v: k for k, v in user_id_map.items()}
item_code_to_id_map = {v: k for k, v in item_id_map.items()}

def recommend_implicit(user_id, model, user_item_matrix_csr, user_id_map, item_code_to_id_map, n_items=10):

    user_code = user_id_map.get(user_id)
    if user_code is None:
        raise ValueError(f"User ID {user_id} not found.")

    recommended, _ = model.recommend(user_code, user_item_matrix_csr[user_code], N=n_items)

    return [item_code_to_id_map.get(item_index, 'Unknown Item') for item_index in recommended]


user_id = df['Username'][0]
recommended_tracks = recommend_implicit(user_id, model_implicit, user_item_matrix_csr, user_id_map, item_code_to_id_map)
print(f"Recommended tracks for user {user_id} using Implicit: {recommended_tracks}")

from implicit.evaluation import precision_at_k, AUC_at_k, ndcg_at_k,mean_average_precision_at_k

# Assuming model, train, and test are already defined
# Calculate precision at k
p_at_k = precision_at_k(model_implicit, train_interaction_matrix, test_interaction_matrix, K=10)

# Calculate AUC
auc_score = AUC_at_k(model_implicit, train_interaction_matrix, test_interaction_matrix)
mean_average_precision_at_k = mean_average_precision_at_k(model_implicit, train_interaction_matrix, test_interaction_matrix)
ndcg_at_k = ndcg_at_k(model_implicit, train_interaction_matrix, test_interaction_matrix)


print(f'Precision at k: {p_at_k}')
print(f'AUC at K: {auc_score}')
print(f'MAP at k: {p_at_k}')
print(f'NDCG at K: {auc_score}')


  0%|          | 0/50 [00:00<?, ?it/s]

Recommended tracks for user emosoup using Implicit: ['A Burning Hill - Mitski', 'A Buried Sun - In Solitude', 'All the Wine - The National', 'Alcohol And Alter Boys - Bayside', 'Alcohol-Free - TWICE', 'Change (In the House of Flies) - Deftones', 'Charles William - All Them Witches', 'A Martyr for My Love for You - The White Stripes', 'All the Ways - Meghan Trainor', 'Addict - N.M.E.']
Recommended tracks for user emosoup using Implicit: ['A Burning Hill - Mitski', 'A Buried Sun - In Solitude', 'All the Wine - The National', 'Alcohol And Alter Boys - Bayside', 'Alcohol-Free - TWICE', 'Change (In the House of Flies) - Deftones', 'Charles William - All Them Witches', 'A Martyr for My Love for You - The White Stripes', 'All the Ways - Meghan Trainor', 'Addict - N.M.E.']


  0%|          | 0/9481 [00:00<?, ?it/s]

  0%|          | 0/9481 [00:00<?, ?it/s]

  0%|          | 0/9481 [00:00<?, ?it/s]

  0%|          | 0/9481 [00:00<?, ?it/s]

  0%|          | 0/9481 [00:00<?, ?it/s]

  0%|          | 0/9481 [00:00<?, ?it/s]

  0%|          | 0/9481 [00:00<?, ?it/s]

  0%|          | 0/9481 [00:00<?, ?it/s]

Precision at k: 0.11132727470613844
AUC at K: 0.5518295140822761
MAP at k: 0.11132727470613844
NDCG at K: 0.5518295140822761
Precision at k: 0.11132727470613844
AUC at K: 0.5518295140822761
MAP at k: 0.11132727470613844
NDCG at K: 0.5518295140822761


# LightFM


In [1]:
import numpy as np
import pandas as pd
from lightfm import LightFM
from lightfm.data import Dataset
from lightfm.evaluation import precision_at_k, auc_score
from lightfm.cross_validation import random_train_test_split
import ast

df = pd.read_csv('/content/drive/MyDrive/Googlecolab/user_songs_filtered.csv')
df['toptags'] = df['toptags'].apply(ast.literal_eval)

# Assuming `df` is your DataFrame with the structure provided.
df_copy = df

In [2]:
dataset = Dataset()
df['track_id'] = df['track_name'] + ' - ' + df['artist_name']
dataset.fit(users=df['Username'].unique(),
            items=df['track_id'].unique(),
            user_features=['country', 'registered_year', "track_count", "artist_count"],
            item_features=['toptags', 'listeners', 'total_playcount', "profanity_density",
                           "polarity", "subjectivity", "emotion1_encoded", "emotion1_score",
                           "emotion2_encoded", "emotion2_score", "mfcc", "chroma", "rms",
                           "spectral_centroid", "zcr", "tempo"])

In [3]:
df['emotion1'].unique()
emotion_mapping = {
    'joy': 1,
    'positive': 2,
    'trust': 3,
    'surprise': 4,
    'negative': 5,
    'sadness': 6,
    'fear': 7,
    'disgust': 8,
    'anger':9
}
df['emotion1_encoded'] = df['emotion1'].map(emotion_mapping)
df['emotion2_encoded'] = df['emotion2'].map(emotion_mapping)
df

Username         country  registered_year  track_count  \
0               emosoup   United States             2015        13520   
1       maiconslavieiro          Brazil             2019        18795   
2             velenious   United States             2013        32367   
3           Antimemetic   United States             2023         4030   
4         frankcreature  Czech Republic             2023         2528   
...                 ...             ...              ...          ...   
393115           Txxinn  United Kingdom             2020        29495   
393116        cow_puppy          Canada             2021        32105   
393117        cow_puppy          Canada             2021        32105   
393118        cow_puppy          Canada             2021        32105   
393119        cow_puppy          Canada             2021        32105   

        artist_count                          track_name  \
0               3386                              Higher   
1               6513                              Higher   
2               8810                              Higher   
3                711                              Higher   
4                706                              Higher   
...              ...                                 ...   
393115          8086                               candy   
393116          2251  Fallen Tree Bridge (Brave Rainbow)   
393117          2251                  The Breathing Fire   
393118          2251       Solar Mirror Anthology Vol. 6   
393119          2251           Outlook Remains Untouched   

                 artist_name  rank  playcount  \
0                Sleep Token     1       1321   
1                Sleep Token    21        151   
2                Sleep Token    19       1259   
3                Sleep Token    32         29   
4                Sleep Token    43         43   
...                      ...   ...        ...   
393115       guccihighwaters    50        334   
393116           Candy Claws    33         39   
393117           Candy Claws    43         35   
393118        Sound of Ceres    46         33   
393119  ||||||||||||||||||||    48         31   

                                                   lyrics  ...  \
0       you say you wont begin again\ncapitulate and l...  ...   
1       you say you wont begin again\ncapitulate and l...  ...   
2       you say you wont begin again\ncapitulate and l...  ...   
3       you say you wont begin again\ncapitulate and l...  ...   
4       you say you wont begin again\ncapitulate and l...  ...   
...                                                   ...  ...   
393115  i might have to block you know its for the bet...  ...   
393116  \nfall asleep up high on moss\ncovered fallen ...  ...   
393117  to halt the freeze of night\ngreen canopy brig...  ...   
393118  \n\nthe king asked her and what was their stor...  ...   
393119  \nhomeostasis refers to the bodys automatic ef...  ...   

        emotion2_score                                               mfcc  \
0             0.156834  [-238.95233    173.60838    -18.518389    40.3...   
1             0.156834  [-238.95233    173.60838    -18.518389    40.3...   
2             0.156834  [-238.95233    173.60838    -18.518389    40.3...   
3             0.156834  [-238.95233    173.60838    -18.518389    40.3...   
4             0.156834  [-238.95233    173.60838    -18.518389    40.3...   
...                ...                                                ...   
393115        0.062549  [-361.05585    255.56912     45.25873    -31.0...   
393116        0.047825  [-55.136433   158.30269    -57.99488     63.88...   
393117        0.139447  [-163.46211     207.27277     -49.669342     1...   
393118        0.123571  [-182.68379    140.46266     -0.2931889   50.1...   
393119        0.158076  [-292.32834    121.62424     43.284298    49.0...   

                                                   chroma       rms  \
0       [0.50727826 0.3796307  0.338861

In [4]:
# Assuming 'dataset' is already initialized and fit with all unique item IDs and feature names
item_features_data = []
for index, row in df.iterrows():
    chroma_list = np.fromstring(row.chroma.strip('[]'), sep=' ')
    chroma_avg = np.mean(chroma_list)
    mfcc_list = np.fromstring(row.mfcc.strip('[]'), sep=' ')
    mfcc_avg = np.mean(mfcc_list)
    # Preparing a dictionary for the current row/item with feature weights
    features_dict = {

        'listeners': int(row.listeners),
        'total_playcount': int(row.total_playcount),
        'profanity_density': float(row.profanity_density),
        'polarity': float(row.polarity),
        'subjectivity': float(row.subjectivity),
        'emotion1_encoded': int(row.emotion1_encoded),  # Convert to integer if it's encoded as a numeric string
        'emotion1_score': float(row.emotion1_score),
        'emotion2_encoded': int(row.emotion2_encoded),  # Convert to integer if it's encoded as a numeric string
        'emotion2_score': float(row.emotion2_score),
        'mfcc': float(mfcc_avg),
        'chroma': float(chroma_avg),
        'rms': float(row.rms),
        'spectral_centroid': float(row.spectral_centroid),
        'zcr': float(row.zcr),
        'tempo': float(row.tempo)
    }

    # Add the item id and its features to the list
    item_features_data.append((row.track_id, features_dict))

# Now, build the item features matrix with this data
item_features = dataset.build_item_features(item_features_data, normalize=True)

In [5]:
# Assuming 'dataset' is already initialized and fit with all unique item IDs and feature names
user_features_data = []
for index, row in df.iterrows():
    # Preparing a dictionary for the current row/item with feature weights
    features_dict = {
        #'country': row.country,
        'registered_year': int(row.registered_year),
        'track_count': int(row.track_count),
        'artist_count': int(row.artist_count)}

    # Add the item id and its features to the list
    item_features_data.append((row.Username, features_dict))

# Now, build the item features matrix with this data
user_features = dataset.build_item_features(user_features_data, normalize=True)

In [6]:

(interactions, weights) = dataset.build_interactions(((row.Username, row.track_id, row.playcount) for index, row in df.iterrows()))

In [9]:
train_interactions, test_interactions = random_train_test_split(interactions, test_percentage=0.2, random_state=None)

# Initialize the LightFM model with your chosen parameters, for example, using the WARP loss function
model = LightFM(loss='warp',learning_rate=0.05, no_components=30)

# Train the model
model.fit(train_interactions, user_features=user_features, item_features=item_features, epochs=50, num_threads=2)

# After training, you can evaluate the model's performance on the test set using LightFM's built-in evaluation metrics
from lightfm.evaluation import precision_at_k, recall_at_k, auc_score

# Compute and print the precision at k
precision_at_k = precision_at_k(model, test_interactions, k=100, user_features=user_features, item_features=item_features).mean()
print(f"Precision at k: {precision_at_k}")
recall_at_k = recall_at_k(model, test_interactions, k=100, user_features=user_features, item_features=item_features).mean()
print(f"recall at k: {recall_at_k}")
auc_score = auc_score(model, test_interactions, user_features=user_features, item_features=item_features).mean()
print(f"auc score: {auc_score}")

# Compute and print the recall at k (requires a custom implementation as LightFM does not provide it directly as of my last update)
# You would need to implement recall at k based on the model predictions and the test interactions



Precision at k: 0.00020050653256475925
recall at k: 0.002668390848592225
auc score: 0.7745291590690613


The AUC score is relatively high, suggesting the model is quite good at distinguishing between songs that a user will like and songs they won’t, when considering the entire dataset.

In the context of a recommendation system using the LightFM model and a user-song interaction matrix, a "relevant song" typically refers to a song that has some form of positive interaction recorded in the interaction matrix for a given user.

epoch: 30 k=5
Precision at k: 0.0002954209630843252
recall at k: 0.00018907612281741844
auc score: 0.7672815918922424

epopch 50, k=100
Precision at k: 0.00016875856090337038
recall at k: 0.0022899736471617073
auc score: 0.7718917727470398

learning rate 0.05, no.components 30
Precision at k: 0.00020050653256475925
recall at k: 0.002668390848592225
auc score: 0.7745291590690613
